cpp/src/strings/copy/repeat_join.cu

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/get_value.cuh>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/strings/copy.hpp>
+#include <cudf/strings/detail/utilities.cuh>
+#include <cudf/strings/detail/utilities.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+
+string_scalar repeat_join(string_scalar const& input,
+                          size_type repeat_times,
+                          rmm::cuda_stream_view stream,
+                          rmm::mr::device_memory_resource* mr)
+{
+  if (!input.is_valid()) { return string_scalar("", false, stream, mr); }
+  if (input.size() == 0 || repeat_times <= 0) { return string_scalar("", true, stream, mr); }
+  if (repeat_times == 1) { return string_scalar(input, stream, mr); }
+
+  CUDF_EXPECTS(input.size() <= std::numeric_limits<size_type>::max() / repeat_times,
+               "The output string has size that exceeds the maximum allowed size.");
+
+  auto buff =
+    rmm::device_buffer(repeat_times * input.size(), stream, rmm::mr::get_current_device_resource());
+  for (size_type i = 0, str_size = input.size(); i < repeat_times; ++i) {
+    auto const offset = i * str_size;
+    thrust::copy(rmm::exec_policy(stream),
+                 input.data(),
+                 input.data() + str_size,
+                 static_cast<char*>(buff.data()) + offset);
+  }
+
+  return string_scalar(string_view(static_cast<char*>(buff.data()), buff.size()), true, stream, mr);
+}
+
+namespace {
+/**
+ * @brief Functor to compute string sizes and repeat the input strings.
+ */
+template <bool has_nulls>
+struct compute_size_and_repeat_fn {
+  column_device_view const strings_dv;
+  size_type repeat_times;
+
+  offset_type* d_offsets{nullptr};
+
+  // If d_chars == nullptr: only compute sizes of the output strings.
+  // If d_chars != nullptr: only repeat strings.
+  char* d_chars{nullptr};
+
+  // Set `1` or `0` for the validities of the output strings if `has_nulls` is `true`.
+  int8_t* d_validities{nullptr};
```

Comment:
Using the validities approach here seems unnecessary inefficient for this API.
The bitmask of the output is identical to the bitmask of the input. 
No need to for this extra memory and extra memory buffer.

Reply:
Ah, yes. Thanks for this.